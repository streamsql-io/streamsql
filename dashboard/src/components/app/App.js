import React from "react";
import { connect } from "react-redux";
import CssBaseline from "@material-ui/core/CssBaseline";
import Box from "@material-ui/core/Box";
import theme from "styles/theme";
import { ThemeProvider } from "@material-ui/core/styles";
import Nav from "components/nav";
import ResourceList from "components/resource-list";
import { Redirect, Route } from "react-router-dom";
import ResourcesAPI from "api/resources";

export const App = ({ sections }) => (
  <ThemeWrapper>
    <Nav sections={sections}>{routes(sections)}</Nav>
  </ThemeWrapper>
);

const apiHandle = new ResourcesAPI();

export const views = {
  RESOURCE_LIST: "ResourceList",
  EMPTY: "Empty",
};

const viewComponent = {
  [views.RESOURCE_LIST]: ResourceList,
  [views.EMPTY]: Box,
};

function routes(sections) {
  const content = parseContentProps(sections);
  const routes = contentRoutes(content);
  const redirect = defaultRedirect(content);
  return routes.concat(redirect);
}

function contentRoutes(content) {
  return content.map((item) => {
    // must be capitalized to work due to:
    // https://reactjs.org/docs/jsx-in-depth.html#choosing-the-type-at-runtime
    const Component = viewComponent[item.view];
    return (
      <Route key={item.path} path={item.path}>
        <Component api={apiHandle} {...item.viewProps} />
      </Route>
    );
  });
}

const defaultRedirect = (content) => {
  const path = indexPath(content);
  // If we just use path as the key, it'll collide with the actual Route, so we
  // prepend default_redirect_.
  const key = "default_redirect_" + path;
  return <Redirect exact key={key} from="/" to={path} />;
};

export const indexPath = (content) => {
  if (content.length === 0) {
    throw TypeError("Nav cannot be empty");
  }
  return content[0].path;
};

export function parseContentProps(sections) {
  return sections
    .flatMap((section) => section.items)
    .filter((item) => !item.external)
    .map(parseItem);
}

const parseItem = (item) => ({
  title: item.title,
  view: item.view || views.EMPTY,
  viewProps: item.viewProps || {},
  path: item.path,
});

export const ThemeWrapper = ({ children }) => (
  <ThemeProvider theme={theme}>
    <CssBaseline />
    {children}
  </ThemeProvider>
);

const mapStateToProps = (state) => ({
  sections: state.navSections,
});

export default connect(mapStateToProps)(App);
