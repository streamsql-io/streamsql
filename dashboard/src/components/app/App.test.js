import React from "react";
// Necessary to get MaterialTable to work correctly.
import "jest-canvas-mock";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { App } from "./App.js";

configure({ adapter: new Adapter() });

describe("App", () => {
  it("hasn't changed", () => {
    expect(shallow(<App />)).toMatchSnapshot();
  });
});
