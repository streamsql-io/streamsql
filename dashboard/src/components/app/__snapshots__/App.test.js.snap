// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App hasn't changed 1`] = `
<Wrapper>
  <Nav
    sections={
      Array [
        Object {
          "items": Array [
            Object {
              "icon": "file-import",
              "path": "/sources",
              "title": "Data Sources",
            },
            Object {
              "icon": "copy",
              "path": "/views",
              "title": "Materialized Views",
            },
            Object {
              "icon": "file-code",
              "path": "/features",
              "title": "Features",
            },
            Object {
              "icon": "sitemap",
              "path": "/feature-sets",
              "title": "Feature Sets",
            },
            Object {
              "icon": "archive",
              "path": "/training-sets",
              "title": "Training Sets",
            },
          ],
          "name": "Resources",
        },
        Object {
          "items": Array [
            Object {
              "icon": "chart-line",
              "path": "/metrics",
              "title": "Metrics",
            },
            Object {
              "icon": "server",
              "path": "/deployment",
              "title": "Deployment",
            },
          ],
          "name": "Monitoring",
        },
        Object {
          "items": Array [
            Object {
              "icon": "users",
              "path": "/users",
              "title": "Users",
            },
            Object {
              "icon": "cogs",
              "path": "/settings",
              "title": "Settings",
            },
            Object {
              "icon": "wallet",
              "path": "/billing",
              "title": "Billing",
            },
            Object {
              "external": true,
              "icon": "book",
              "path": "https://docs.streamsql.io",
              "title": "Documentation",
            },
            Object {
              "icon": "question",
              "path": "/help",
              "title": "Help",
            },
          ],
          "name": "Admin",
        },
      ]
    }
  >
    <ResourceList />
  </Nav>
</Wrapper>
`;
